### PR TITLE
Add Apache Spark 3.1.4-SNAPSHOT Shims

### PR DESCRIPTION
**Pull request by nartal1**
This fixes https://github.com/NVIDIA/spark-rapids/issues/4789.

This PR adds Spark-3.1.4-SNAPSHOT shims. Added test to verify that it works as expected.

### CODE REVIEW

**Comment by nartal1** [comment]
@tgravescs Could you please let me know the priority of this over https://github.com/NVIDIA/spark-rapids/pull/4857. I can wait until that PR is merged and upmerge my changes to reflect the new design. If we want to enable 3.1.4-SNAPSHOT nightly builds sooner then I think this should be merged earlier.

**Comment by tgravescs** [comment]
either way, I think 3.1.4 can wait if the other PR is large

**Comment by nartal1** [comment]
> either way, I think 3.1.4 can wait if the other PR is large

Thanks for the confirmation. Have marked this PR as draft. Will upmerge once other PR is merged.

**Comment by razajafri** [comment]
Thanks @nartal1

**Comment by tgravescs** [comment]
@nartal1 please update, the other pr has been merged

File: docs/additional-functionality/rapids-shuffle.md [hunk 1]
```diff
@@ -285,19 +285,20 @@ In this section, we are using a docker container built using the sample dockerfi
...
+    | Spark Shim     | spark.shuffle.manager value                              |
+----------------| --------------| -------------------------------------------------------- |
+    | 3.0.1          | com.nvidia.spark.rapids.spark301.RapidsShuffleManager    |
+    | 3.0.2          | com.nvidia.spark.rapids.spark302.RapidsShuffleManager    |
+    | 3.0.3          | com.nvidia.spark.rapids.spark303.RapidsShuffleManager    |
+    | 3.1.1          | com.nvidia.spark.rapids.spark311.RapidsShuffleManager    |
+    | 3.1.1 CDH      | com.nvidia.spark.rapids.spark311cdh.RapidsShuffleManager |
+    | 3.1.2          | com.nvidia.spark.rapids.spark312.RapidsShuffleManager    |
+    | 3.1.3          | com.nvidia.spark.rapids.spark313.RapidsShuffleManager    |
+    | 3.1.4          | com.nvidia.spark.rapids.spark314.RapidsShuffleManager    |
```
Comment (tgravescs) [inline docs/additional-functionality/rapids-shuffle.md:297]: leave this off until it becomes not a snapshot

File: jenkins/Jenkinsfile-blossom.premerge [hunk 2]
```diff
@@ -328,7 +328,7 @@ pipeline {
                                                                       'shims/spark301db/src/main/scala/,shims/spark301/src/main/scala/,' +
                                                                       'shims/spark302/src/main/scala/,shims/spark303/src/main/scala/,' +
                                                                       'shims/spark304/src/main/scala/,shims/spark312/src/main/scala/,' +
-                                                                      'shims/spark313/src/main/scala/'
+                                                                      'shims/spark313/src/main/scala/,shims/spark314/src/main/scala/'
```
Comment (tgravescs) [inline jenkins/Jenkinsfile-blossom.premerge:331]: this directory doesn't exist anymore, revert

Comment (nartal1) [reply jenkins/Jenkinsfile-blossom.premerge]: done

File: sql-plugin/src/test/314/scala/com/nvidia/spark/rapids/shims/spark314/Spark314ShimsSuite.scala [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import org.scalatest.FunSuite
+
+class Spark314ShimsSuite extends FunSuite {
+  test("spark shims version") {
+    assert(SparkShimImpl.getSparkShimVersion === SparkShimVersion(3, 1, 4))
+  }
+
+  test("shuffle manager class") {
+    assert(ShimLoader.getRapidsShuffleManagerClass ===
+      classOf[com.nvidia.spark.rapids.spark314.RapidsShuffleManager].getCanonicalName)
```
[the remark this answers is not in the document]
Comment (nartal1) [reply sql-plugin/src/test/314/scala/com/nvidia/spark/rapids/shims/spark314/Spark314ShimsSuite.scala]: removed these tests.

**Comment by tgravescs** [comment]
build

**Review by tgravescs — APPROVED** [review]
We need to get integration test pipeline going for this.

**Pull request merged**